### PR TITLE
support Avx2

### PR DESCRIPTION
**Pull request by yuchengliu1**
## Type of Change

feature
No API changed

Run LLM on client CPU with AVX2 (without avx512)

detail description 
JIRA ticket: xxx

## Expected Behavior & Potential Risk
CPU: i7-9850H@2.6GHz 
memory: single channel 32GB@2666MHz (memory bandwidth 21.3GB/s)
compute_type is FP32
<html>
<body>
<span><span class="ui-provider bwq bwr azv bws bwt bwu bwv bww bwx bwy bwz bxa bxb bxc bxd bxe bxf bxg bxh bxi bxj bxk bxl bxm bxn bxo bxp bxq bxr bxs bxt bxu bxv bxw bxx" dir="ltr"><span><p>&nbsp;</p>

  | first token | next token
-- | -- | --
gptj-6B_q4j_b128 | 382.57 ms (63.76 ms per token) | 220.18 ms per token
llama-7B_q4j_b128 | 515.35 ms (73.62 ms per token) | 250.71 ms per token
llama2-7B_q4j_b128 | 518.14 ms (74.02 ms per token) | 253.23 ms per token

<p>&nbsp;</p></span></span></span>
</body>
</html>

## How has this PR been tested?

how to reproduce the test (including hardware information)

## Dependency Change?

any library dependency introduced or removed

### CODE REVIEW

File: intel_extension_for_transformers/llm/runtime/graph/vectors/cpu/vec_base.hpp [hunk 1]
```diff
@@ -54,16 +68,49 @@ typedef std::pair<__m256i, __m256i> int32x16;
...
+  T* dst_T = reinterpret_cast<T*>(dst);
+  *dst_T = src;
+}
+
+template <>
+void store_kernel_t<s8x16>(void* dst, s8x16 src);
+template <>
+void store_kernel_t<fp32x16>(void* dst, fp32x16 src);
+template <>
+void store_kernel_t<bf16x16>(void* dst, bf16x16 src);
```
Comment (DDEle) [inline intel_extension_for_transformers/llm/runtime/graph/vectors/cpu/vec_base.hpp:113]: There seems to be duplicated definitions here.
image

**Comment by airMeng** [comment]
I think you already remove all warnings in this PR?

**Comment by airMeng** [comment]
> test on [i7-9850H@2.6Hz](mailto:i7-9850H@2.6Hz)
> 
>  
> 
>  	first token	next token
> gptj-6B_q4j_b128	382.57 ms (63.76 ms per token)	220.18 ms per token
> llama-7B_q4j_b128	515.35 ms (73.62 ms per token)	250.71 ms per token
> llama2-7B_q4j_b128	518.14 ms (74.02 ms per token)	253.23 ms per token
> 


Can you provide memory bandwidth comparison between clients and SPR?  It's useful to judge whether the gap is meaningful.

File: intel_extension_for_transformers/llm/runtime/graph/README.md [hunk 2]
```diff
@@ -61,7 +61,7 @@ cd build
 cmake ..
 cmake --build . -j
 ```
-
+Note: add compile args ```-DNE_AVX512=OFF -DNE_AVX512_VBMI=OFF -DNE_AVX512_VNNI=OFF``` to ```cmake``` when compiling it on a CPU without AVX512
```
Comment (a32543254) [inline intel_extension_for_transformers/llm/runtime/graph/README.md:64]: could we auto detect the machine's isa to add follow compile args without manually ?

File: intel_extension_for_transformers/llm/runtime/graph/vectors/cpu/vec_load.cpp [hunk 3]
```diff
@@ -13,38 +13,3 @@
 //  limitations under the License.
 
 #include "vec_load.hpp"
```
Comment (a32543254) [inline intel_extension_for_transformers/llm/runtime/graph/vectors/cpu/vec_load.cpp:15]: this vec_load.cpp become empty file for now, do we still need keep this cpp file ?

File: intel_extension_for_transformers/llm/runtime/graph/models/model_utils/model_utils.cpp [hunk 4]
```diff
@@ -926,7 +926,7 @@ size_t jblas_quantize(const float* f32ptr, void* dstpr, const quant_params_inter
         using KernelRef = WeiS8Fp32<GcCompFp32, JblasNoSIMD>;
         static Kernel kernel;
         static KernelRef kernelref;
-        auto packedw = kernel.createStorage(n, k, params.group_size, params.alg == quant_alg::sym);
+        auto packedw = kernel.createStorage(n, k, params.group_size, params.alg == quant_alg::asym);
```
Comment (a32543254) [inline intel_extension_for_transformers/llm/runtime/graph/models/model_utils/model_utils.cpp:929]: why change sym to asym ? 
the function is still same ?

File: intel_extension_for_transformers/llm/runtime/graph/vectors/cpu/vec_base.hpp (hunk 1 above)
Comment (yuchengliu1) [reply intel_extension_for_transformers/llm/runtime/graph/vectors/cpu/vec_base.hpp:113]: No duplication here. Line 103 is the declarations and definitions for the template function, and it has some specializations for some type. The declaration of specializations are in vec_base.hpp while their definitions are in vec_store.hpp.

File: intel_extension_for_transformers/llm/runtime/graph/vectors/cpu/vec_compare.hpp [hunk 5]
```diff
@@ -17,13 +17,13 @@
 
 #include "vec_base.hpp"
 
-inline fp32x16 min_fp32x16(fp32x16 a, fp32x16 b);
+fp32x16 min_fp32x16(fp32x16 a, fp32x16 b);
```
[the remark this answers is not in the document]
Comment (yuchengliu1) [reply intel_extension_for_transformers/llm/runtime/graph/vectors/cpu/vec_compare.hpp:20]: A inline function should declare and define in `.hpp` .  We want to compile some `.cpp` to a static library. Thus, inline function cannot be used here.

File: intel_extension_for_transformers/llm/runtime/graph/vectors/cpu/vec_load.cpp (hunk 3 above)
Comment (yuchengliu1) [reply intel_extension_for_transformers/llm/runtime/graph/vectors/cpu/vec_load.cpp:15]: Yes, we should compile a static library by `.cpp`

File: intel_extension_for_transformers/llm/runtime/graph/models/model_utils/model_utils.cpp (hunk 4 above)
Comment (yuchengliu1) [reply intel_extension_for_transformers/llm/runtime/graph/models/model_utils/model_utils.cpp:929]: This code is wrong before. This PR fix it.

**Comment by yuchengliu1** [comment]
> I think you already remove all warnings in this PR?

Yes, there is no warning now.

**Review by airMeng — APPROVED** [review]

**Review by a32543254 — APPROVED** [review]
LGTM

**Comment by a32543254** [comment]
Better we also give the performance between ITREX AVX2 and GGML AVX2 to find our benefit or gap.

File: intel_extension_for_transformers/llm/runtime/graph/vectors/cpu/vec_base.hpp (hunk 1 above)
Comment (DDEle) [reply intel_extension_for_transformers/llm/runtime/graph/vectors/cpu/vec_base.hpp:113]: You are right, but MSVC tolds me that there are duplications about store_kernel_t.
``` bat
C:\itrex\intel_extension_for_transformers\llm\runtime\graph\out\build\x64-release-avx2\vec_convert.cpp.obj : warning LNK4006: "void __cdecl store_kernel_t<struct s8x16>(void *,struct s8x16)" (??$store_kernel_t@Us8x16@@@@YAXPEAXUs8x16@@@Z) already defined in vec_store.cpp.obj; second definition ignored
C:\itrex\intel_extension_for_transformers\llm\runtime\graph\out\build\x64-release-avx2\vec_convert.cpp.obj : warning LNK4006: "void __cdecl store_kernel_t<struct fp32x16>(void *,struct fp32x16)" (??$store_kernel_t@Ufp32x16@@@@YAXPEAXUfp32x16@@@Z) already defined in vec_store.cpp.obj; second definition ignored
C:\itrex\intel_extension_for_transformers\llm\runtime\graph\out\build\x64-release-avx2\vec_convert.cpp.obj : warning LNK4006: "void __cdecl store_kernel_t<struct bf16x16>(void *,struct bf16x16)" (??$store_kernel_t@Ubf16x16@@@@YAXPEAXUbf16x16@@@Z) already defined in vec_store.cpp.obj; second definition ignored
C:\itrex\intel_extension_for_transformers\llm\runtime\graph\out\build\x64-release-avx2\vec_arithmetic.cpp.obj : warning LNK4006: "struct fp32x16 __cdecl load_kernel_t<struct fp32x16>(void const *)" (??$load_kernel_t@Ufp32x16@@@@YA?AUfp32x16@@PEBX@Z) already defined in vec_load.cpp.obj; second definition ignored
C:\itrex\intel_extension_for_transformers\llm\runtime\graph\out\build\x64-release-avx2\vec_arithmetic.cpp.obj : warning LNK4006: "struct bf16x16 __cdecl load_kernel_t<struct bf16x16>(void const *)" (??$load_kernel_t@Ubf16x16@@@@YA?AUbf16x16@@PEBX@Z) already defined in vec_load.cpp.obj; second definition ignored
C:\itrex\intel_extension_for_transformers\llm\runtime\graph\out\build\x64-release-avx2\vec_arithmetic.cpp.obj : warning LNK4006: "void __cdecl store_kernel_t<struct s8x16>(void *,struct s8x16)" (??$store_kernel_t@Us8x16@@@@YAXPEAXUs8x16@@@Z) already defined in vec_store.cpp.obj; second definition ignored
C:\itrex\intel_extension_for_transformers\llm\runtime\graph\out\build\x64-release-avx2\vec_arithmetic.cpp.obj : warning LNK4006: "void __cdecl store_kernel_t<struct fp32x16>(void *,struct fp32x16)" (??$store_kernel_t@Ufp32x16@@@@YAXPEAXUfp32x16@@@Z) already defined in vec_store.cpp.obj; second definition ignored
  
C:\itrex\intel_extension_for_transformers\llm\runtime\graph\out\build\x64-release-avx2\vec_arithmetic.cpp.obj : warning LNK4006: "void __cdecl store_kernel_t<struct bf16x16>(void *,struct bf16x16)" (??$store_kernel_t@Ubf16x16@@@@YAXPEAXUbf16x16@@@Z) already defined in vec_store.cpp.obj; second definition ignored
```

**Comment by yuchengliu1** [comment]
> Better we also give the same models's performance between ITREX AVX2 and GGML AVX2 to find our benefit or gap.

<html xmlns:v="urn:schemas-microsoft-com:vml"
xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:x="urn:schemas-microsoft-com:office:excel"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=Excel.Sheet>
<meta name=Generator content="Microsoft Excel 15">

</head>

</span>

<body link="#0563C1" vlink="#954F72">



llama2-7B  | first token | next token
-- | -- | --
llama.cpp | 494.15 ms (61.77 ms per token) | 141.69 ms per token
ITREX   q4_0 | 743.89ms (106.27ms per token) | 136.37ms per token
ITREX   q4_j | 518.14 ms (74.02 ms per token) | 253.23 ms per token



</body>

</html>

the profiling of ITREX AVX2 (profile on q4_j) 
perf_total_per_op_us[                     ADD] =   0.184 ms
perf_total_per_op_us[                     MUL] =   0.327 ms
perf_total_per_op_us[                    SILU] =   3.488 ms
perf_total_per_op_us[                RMS_NORM] =   0.396 ms
perf_total_per_op_us[                 MUL_MAT] =   2.497 ms
perf_total_per_op_us[                   SCALE] =   0.032 ms
perf_total_per_op_us[                     CPY] =   1.177 ms
perf_total_per_op_us[                 RESHAPE] =   0.037 ms
perf_total_per_op_us[                    VIEW] =   0.066 ms
perf_total_per_op_us[                 PERMUTE] =   0.029 ms
perf_total_per_op_us[               TRANSPOSE] =   0.006 ms
perf_total_per_op_us[                GET_ROWS] =   0.008 ms
perf_total_per_op_us[           DIAG_MASK_INF] =   0.025 ms
perf_total_per_op_us[                SOFT_MAX] =   0.166 ms
perf_total_per_op_us[                    ROPE] =   2.940 ms
perf_total_per_op_us[           INNER PRODUCT] = 254.974 ms
========================================

the profiling of GGML AVX2 (profile on q4_0)
 perf_total_per_op_us[                     ADD] =   0.119 ms
perf_total_per_op_us[                     MUL] =   0.244 ms
perf_total_per_op_us[                    SILU] =   3.395 ms
perf_total_per_op_us[                RMS_NORM] =   0.345 ms
perf_total_per_op_us[                 MUL_MAT] =   2.281 ms
perf_total_per_op_us[                   SCALE] =   0.024 ms
perf_total_per_op_us[                     CPY] =   0.976 ms
perf_total_per_op_us[                 RESHAPE] =   0.032 ms
perf_total_per_op_us[                    VIEW] =   0.063 ms
perf_total_per_op_us[                 PERMUTE] =   0.023 ms
perf_total_per_op_us[               TRANSPOSE] =   0.017 ms
perf_total_per_op_us[                GET_ROWS] =   0.004 ms
perf_total_per_op_us[           DIAG_MASK_INF] =   0.015 ms
perf_total_per_op_us[                SOFT_MAX] =   0.117 ms
perf_total_per_op_us[                    ROPE] =   0.271 ms
perf_total_per_op_us[           INNER PRODUCT] = 123.976 ms
========================================

File: intel_extension_for_transformers/llm/runtime/graph/vectors/cpu/vec_base.hpp (hunk 1 above)
Comment (yuchengliu1) [reply intel_extension_for_transformers/llm/runtime/graph/vectors/cpu/vec_base.hpp:113]: fix in [PR511](https://github.com/intel/intel-extension-for-transformers/pull/511)

File: intel_extension_for_transformers/llm/runtime/graph/README.md (hunk 2 above)
Comment (yuchengliu1) [reply intel_extension_for_transformers/llm/runtime/graph/README.md:64]: fix in [PR511](https://github.com/intel/intel-extension-for-transformers/pull/511)

Comment (DDEle) [reply intel_extension_for_transformers/llm/runtime/graph/README.md:64]: I don't think so. If I remember correctly, there was a similar discussion for the deprecated executor, and the result is that it is not possible to find the hardware info of the runtime match during compile time on the machine that compiles the executable.
cc @luoyu-intel

**Pull request merged**